### PR TITLE
Ptref: use map for uri look up

### PR DESCRIPTION
**Pull request by antoine-de**
for uri look up we don't use the vector anymore, we use the associated maps.

This should speed up 'a bit' the ptref look up

[Review on Reviewable](https://reviewable.io/reviews/canaltp/navitia/1435)

### CODE REVIEW

**Comment by xlqian** [comment]
_[source/jormungandr/tests/ptref_tests.py, line 337 [r1]](https://reviewable.io:443/reviews/canaltp/navitia/1435#-KBgFtLpwow4R-hCrRLd:-KBgFtLpwow4R-hCrRLe:19162292) ([raw file](https://github.com/canaltp/navitia/blob/4f84eb1682f6d468e31aedfadcafc914991e5ce6/source/jormungandr/tests/ptref_tests.py#L337)):_
why is this changed?

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canaltp/navitia/1435)_

**Comment by antoine-de** [comment]
Review status: 0 of 17 files reviewed at latest revision, 1 unresolved discussion.

---

_[source/jormungandr/tests/ptref_tests.py, line 337 [r1]](https://reviewable.io:443/reviews/canaltp/navitia/1435#-KBgFtLpwow4R-hCrRLd:-KBgHA4iezJaw_06F6Xx:-1268636774) ([raw file](https://github.com/canaltp/navitia/blob/4f84eb1682f6d468e31aedfadcafc914991e5ce6/source/jormungandr/tests/ptref_tests.py#L337)):_
because the maps were not correctly used in the builder :wink: so in this case there were 2 physical_mode::Car ...

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canaltp/navitia/1435)_

**Comment by xlqian** [comment]
_[source/ptreferential/ptreferential.cpp, line 201 [r1]](https://reviewable.io:443/reviews/canaltp/navitia/1435#-KBgJV83E263WjuUYcKp:-KBgJV84eS3sRbguJnYP:-978735724) ([raw file](https://github.com/canaltp/navitia/blob/4f84eb1682f6d468e31aedfadcafc914991e5ce6/source/ptreferential/ptreferential.cpp#L201)):_
An alternative solution: https://ideone.com/KmI3Sb 

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canaltp/navitia/1435)_

**Comment by antoine-de** [comment]
Review status: 0 of 17 files reviewed at latest revision, 2 unresolved discussions.

---

_[source/ptreferential/ptreferential.cpp, line 201 [r1]](https://reviewable.io:443/reviews/canaltp/navitia/1435#-KBgJV83E263WjuUYcKp:-KBgNBJBaGApaSzBdY9z:-1413446551) ([raw file](https://github.com/canaltp/navitia/blob/4f84eb1682f6d468e31aedfadcafc914991e5ce6/source/ptreferential/ptreferential.cpp#L201)):_
it's not possible 'as is' as ObjFactory has no `key_type`.
we can add it (and It would be a good thing), but I find that it's more straightforward with the decltype than with `key_type`)

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canaltp/navitia/1435)_

**Comment by xlqian** [comment]
ok for me

File: source/ptreferential/ptreferential.cpp [hunk 1]
```diff
@@ -186,6 +186,40 @@ std::vector<idx_t> filtered_indexes(const T& data, const C& clause) {
...
+    return std::true_type{};
+}
+template <typename Container>
+inline constexpr std::false_type has_string_lookup(const Container&, ...) {
+    return std::false_type{};
+}
+
+template<typename T, typename Container>
+std::vector<idx_t> filter_by_uri(const Container& data, const std::string& uri, std::false_type) {
+    // by refault, we loop over all element to find the one
```
Comment (TeXitoi) [inline source/ptreferential/ptreferential.cpp]: by **d**efault

**Comment by TeXitoi** [comment]
OK for me.

**Pull request merged**